spec.html

```diff
@@ -42406,88 +42519,111 @@ <h1>
           1. Let _eventsRecord_ be the Agent Events Record of _execution_.[[EventsRecords]] whose [[AgentSignifier]] is AgentSignifier().
           1. Let _leaveEvent_ be a new Synchronize event.
           1. Append _leaveEvent_ to _eventsRecord_.[[EventList]].
-          1. Set the Synchronize event in _WL_ to _leaveEvent_.
+          1. Set _WL_.[[MostRecentLeaveEvent]] to _leaveEvent_.
           1. Leave the critical section for _WL_.
           1. Return ~unused~.
         </emu-alg>
       </emu-clause>
 
+      <emu-clause id="sec-triggertimeout" type="abstract operation">
+        <h1>
+          TriggerTimeout (
+            _WL_: a WaiterList Record,
+            _waiterRecord_: a Waiter Record
+          ): ~unused~
+        </h1>
+        <dl class="header">
+        </dl>
+        <emu-alg>
+          1. Assert: _waiterRecord_.[[Timeout]] is finite.
+          1. Perform EnterCriticalSection(_WL_).
+          1. If _waiterRecord_ is in _WL_.[[Waiters]], then
+            1. Set _waiterRecord_.[[Result]] to *"timed-out"*.
+            1. Perform RemoveWaiter(_WL_, _waiterRecord_).
+            1. Perform NotifyWaiter(_WL_, _waiterRecord_).
+          1. Perform LeaveCriticalSection(_WL_).
+          1. Return ~unused~.
+        </emu-alg>
+      </emu-clause>
+
       <emu-clause id="sec-addwaiter" type="abstract operation">
         <h1>
           AddWaiter (
-            _WL_: a WaiterList,
-            _W_: an agent signifier,
+            _WL_: a WaiterList Record,
+            _waiterRecord_: a Waiter Record,
           ): ~unused~
         </h1>
         <dl class="header">
         </dl>
         <emu-alg>
           1. Assert: The surrounding agent is in the critical section for _WL_.
-          1. Assert: _W_ is not on the list of waiters in any WaiterList.
-          1. Append _W_ to _WL_.
+          1. Assert: There is no Waiter Record in _WL_.[[Waiters]] whose [[PromiseCapability]] field is _waiterRecord_.[[PromiseCapability]] and [[AgentSignifier]] field is _waiterRecord_.[[AgentSignifier]].
+          1. Append _waiterRecord_ as the last element of _WL_.[[Waiters]].
+          1. Let _minimumTimeout_ be _waiterRecord_.[[Timeout]].
+          1. If _minimumTimeout_ is finite, then in parallel
```

Comment:
The definition of "in parallel" distinguishes between (a) "a step that ends with 'in parallel'" and (b) "subsequent steps in the algorithm". Here, the steps 5.a - 5.d come after the step that ends with 'in parallel', and yet they need to be considered part of (a), not (b).

It might be cleaner to put the parallel steps in an abstract closure, and then say something like:
```
Perform _waitAndTimeout_() in parallel.
```
or
```
Spawn _waitAndTimeout_().
```

(And then, instead of `in parallel`, you'd define `Perform ... in parallel` or `Spawn`, and not have to handle "if a parallel step is subdivided".)

Reply:
Removed parallel steps entirely.